R/preprocessing.R

```diff
@@ -0,0 +1,152 @@
+##'   zero. If "zero" then all observations on missing dates will be assumed to
+##'   be of value 0.
+##' @param na Character. How to process dates that have NA values. The options
+##'   available are the same ones as for the `missing` argument.
+##' @param initial_accumulate Integer. The number of initial dates to accumulate
+##'   if `missing` or a column name is set to `"accumulate"`. This number of
+##'   dates is added to the beginning of the data set to be accumulated onto the
+##'   first data point. This is useful, for example, for modelling weekly
+##'   incidence data, in which case this should be set to 7. If accumulating and
+##'   the first data point is one to accumulate and this is not set, then that
+##'   data point will be removed with a warning.
+##' @param obs_column Character. If given, only the column specified here will
+##'   be used for checking missingness. This is useful if using a data set that
+##'   has NA values in multiple columns, but only one of them corresponds to
+##'   observations that are to be processed here. By default, all columns will
+##'   be processed and, e.g. `accumulate` will be set if any column is NA in any
+##'   given row.
+##' @return a data.table with an `accumulate` column that indicates whether
+##'   values are accumulated (see the documnetation of the `data` arugment in
+##'   [estimate_infections()])
+##' @importFrom rlang arg_match
+##' @importFrom data.table as.data.table data.table
+##' @export
+##' @examples
+##' cases <- data.table::copy(example_confirmed)
+##' ## calculate weekly sum
+##' cases[, confirm := data.table::frollsum(confirm, 7)]
+##' ## limit to dates once a week
+##' cases <- cases[seq(7, nrow(cases), 7)]
+##' ## set the second observation to missing
+##' cases[2, confirm := NA]
+##' ## fill missing data
+##' fill_missing(cases, missing = "accumulate", initial_accumulate = 7)
+fill_missing <- function(data,
+                         missing = c("missing", "accumulate", "zero"),
+                         na = c("missing", "accumulate", "zero"),
+                         initial_accumulate,
+                         obs_column) {
+  assert_data_frame(data)
+  assert_names(names(data), must.include = "date", disjunct.from = "accumulate")
+  if (!missing(obs_column)) {
+    assert_names(names(data), must.include = obs_column)
+  } else {
+    obs_column <- setdiff(colnames(data), "date")
+  }
+  assert_date(data$date, any.missing = FALSE)
+
+  data <- as.data.table(data)
+
+  missing <- arg_match(missing)
+  na <- arg_match(na)
+
+  ## first, processing missing dates
+  initial_add <- ifelse(missing(initial_accumulate), 1, initial_accumulate)
+
+  date_seq <- seq.Date(
+    min(data$date) - initial_add + 1,
+    max(data$date), by = "day"
+  )
+
+  ## mark dates that are present in the data
+  data[, ..present := TRUE]
+  complete_dates <- data.table(date = date_seq)
+  data <- merge(data, complete_dates, by = "date", all.y = TRUE)
```

Comment:
if people are preparing data from multi-regions will this break? does it matter if it does

Reply:
I hadn't considered this. We could add a `by` argument, scoringutils-style?